tts/src/tts/synthesizer.py

```diff
@@ -103,22 +174,81 @@ def __init__(self, engine='POLLY_SERVICE', polly_service_name='polly'):
-            tmp_filepath = os.path.join(os.sep, 'tmp', 'voice_{}_{}'.format(tmp_filename, str(time.time())))
+            tmp_filename = hashlib.md5(
+                json.dumps(kw, sort_keys=True)).hexdigest()
+            tmp_filepath = os.path.join(
+                os.sep, 'tmp', 'voice_{}'.format(tmp_filename))
             kw['output_path'] = os.path.abspath(tmp_filepath)
-        rospy.loginfo('audio will be saved as {}'.format(kw['output_path']))
-
-        return self.engine(**kw)
+            rospy.loginfo('managing file with name: {}'.format(tmp_filename))
+
+            # because the hash will include information about any file ending choices, we only
+            # need to look at the hash itself.
+            db = DB()
+            db_search_result = db.ex(
+                'SELECT file, audio_type FROM cache WHERE hash=?', tmp_filename).fetchone()
+            current_time = time.time()
+            file_found = False
+            if db_search_result:  # then there is data
+                # check if the file exists, if not, remove from db
+                # TODO: add a test that deletes a file without telling the db and tries to synthesize it
+                if os.path.exists(db_search_result['file']):
+                    file_found = True
+                    db.ex('update  cache set last_accessed=? where hash=?',
+                          current_time, tmp_filename)
+                    synth_result = PollyResponse(json.dumps({
+                        'Audio File': db_search_result['file'],
+                        'Audio Type': db_search_result['audio_type'],
+                        'Amazon Polly Response Metadata': ''
+                    }))
+                    rospy.loginfo('audio file was already cached at: %s',
+                                  db_search_result['file'])
+                else:
+                    rospy.logwarn(
+                        'A file in the database did not exist on the disk, removing from db')
+                    db.remove_file(db_search_result['file'])
+            if not file_found:  # havent cached this yet
+                rospy.loginfo('Caching file')
+                synth_result = self.engine(**kw)
+                res_dict = json.loads(synth_result.result)
+                if 'Exception' not in res_dict:
+                    file_name = res_dict['Audio File']
+                    if file_name:
+                        file_size = os.path.getsize(file_name)
+                        db.ex('''insert into cache(
+                            hash, file, audio_type, last_accessed,size)
+                            values (?,?,?,?,?)''', tmp_filename, file_name,
+                              res_dict['Audio Type'], current_time, file_size)
+                        rospy.loginfo(
+                            'generated new file, saved to %s and cached', file_name)
+                        # make sure the cache hasn't grown too big
```

Comment:
#### Performance concerns
* Potentially doing many more transactions than needed; only need to do at most one transaction per _call_engine call. For example, if the new file caused us to go beyond the size limit by 1000 bytes, and we need to remove the 10 oldest files in order to clear that much space, we'd be doing more than 10 transactions.
* Seems like everything is synchronous, so even if we only have one transaction (say, to insert a new entry) - returning `synth_result` is still blocked on completion of that transaction. Using WAL and performing checkpointing in a separate thread is one way to overcome that.

#### Performance Impact discussion
In part, the value of this node is in enabling Human-Robot interaction via Polly, and that heavily depends on the latency; if I'm asking the robot a question and it answers 5 seconds later, that's not very useful. 

We know that with this implementation there will be increased latency, the question is how much. Would it add 1ms, or 500ms? maybe it's even less than 1ms, I really don't know. Do you? 
It would feel a bit premature to me to merge this without doing a simple benchmark in which we mock out line 228 (the self.engine() call) and measure the added latency (in lieu of any other data).

#### Path forward
Generally, it seems like we're reinventing the wheel here by trying to use SQLite directly. There are dozens of libraries that implemented all the convenience functionality we'd ever need (see [DiskCache](http://www.grantjenks.com/docs/diskcache/tutorial.html#cache)) as well as performance / transaction optimizations. So to summarize:
* Why not use some of the battle tested disk-backed caches out there? Did you choose sqlite because of lack of integration with ROS? (i.e. not available via rosdep?)
* If we have to continue using SQLite directly, we need to test the effect on latency. Again, maybe there's no issue at all, but we need to make sure.

Reply:
Sorry I have been so slow to respond. I'm going to start here before going anywhere else, since the answers on this thread may change what is being done elsewhere.

I totally agree that latency is a major issue in these types of systems. SQLite is fast and I would imagine that it is at least an order of magnitude faster than making a web call that downloads a file, probably multiple orders of magnitude better. But you are correct, I don't know that for a fact. 

I chose SQLite because it is built into python, it is stable, and it is well known/documented. I could certainly see another tool being used, but I don't know which other tools would check those boxes. DiskCache that you link to looks great. Is that a dependency you all are ok with?

Reply:
Hey @AAlon, want to check back in on this. If you would like me to refactor this to DiskCache or make some timing tests for the current implementation, I can do either of those. DiskCache would make for cleaner more maintainable code, but not sure if that is a dependency that y'all want in here. Just let me know, I'm hoping to get some code done in a few places over the next few weeks.

Reply:
I think that's fine, but we should probably take a look at the benchmarking mentioned previously.

Reply:
what is "that" in "that's fine"?

Reply:
Sorry, I'll be specific. @mjsobrep if you would like to implement some timing tests, in order to provide a ground truth, and test your changes here (along with possibly refactoring to DiskCache) I think that's a good start. 

I'm not sure if you'e seen it, but the [DiskCache timing info](http://www.grantjenks.com/docs/diskcache/#comparisons) seems promising.